% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/spca.R
\name{spca}
\alias{spca}
\alias{spca.default}
\alias{spca.character}
\title{Sparse Principal Components Analysis}
\usage{
spca(X = NULL, data = NULL, ncomp = 2, keepX = NULL, ...)

\method{spca}{default}(X, ncomp = 2, center = TRUE, scale = TRUE,
  keepX = NULL, max.iter = 500, tol = 1e-06, logratio = c("none",
  "CLR"), multilevel = NULL)

\method{spca}{character}(X = NULL, data = NULL, ncomp = 2,
  keepX = NULL, ...)
}
\arguments{
\item{X}{A numeric matrix (or data frame) which provides the data for the
principal components analysis. It can contain missing values.
Alternatively, name of an assay from \code{data}.}

\item{data}{An object of class \code{MultiAssayExperiment} or similar. See Details.}

\item{ncomp}{Integer, if data is complete \code{ncomp} decides the number of
components and associated eigenvalues to display from the \code{pcasvd}
algorithm and if the data has missing values, \code{ncomp} gives the number
of components to keep to perform the reconstitution of the data using the
NIPALS algorithm. If \code{NULL}, function sets \code{ncomp = min(nrow(X),
ncol(X))}.}

\item{keepX}{numeric vector of length ncomp, the number of variables to keep.}

\item{...}{Aguments passed to the generic.}

\item{center}{A logical value indicating whether the variables should be
shifted to be zero centered. Alternately, a vector of length equal the
number of columns of \code{X} can be supplied. The value is passed to
\code{\link{scale}}.}

\item{scale}{A logical value indicating whether the variables should be
scaled to have unit variance before the analysis takes place. The default is
\code{FALSE} for consistency with \code{prcomp} function, but in general
scaling is advisable. Alternatively, a vector of length equal the number of
columns of \code{X} can be supplied. The value is passed to
\code{\link{scale}}.}

\item{max.iter}{Integer, the maximum number of iterations in the NIPALS
algorithm.}

\item{tol}{A positive real, the tolerance used in the NIPALS algorithm.}

\item{logratio}{one of ('none','CLR'). Specifies the log ratio
transformation to deal with compositional values that may arise from
specific normalisation in sequencing data. Default to 'none'.
in loading vectors. By default all variables are kept in the model. See
details.}

\item{multilevel}{Sample information for multilevel decomposition for repeated measurements.}
}
\value{
\code{spca} returns a list with class \code{"spca"} containing the
following components: \item{ncomp}{the number of components to keep in the
calculation.} \item{varX}{the adjusted cumulative percentage of variances
explained.} \item{keepX}{the number of variables kept in each loading
vector.} \item{iter}{the number of iterations needed to reach convergence
for each component.} \item{rotation}{the matrix containing the sparse
loading vectors.} \item{x}{the matrix containing the principal components.}
Refer to \code{pca} for different type of inputs supported.
}
\description{
Performs a sparse principal components analysis to perform variable
selection by using singular value decomposition.

The calculation employs singular value decomposition of the (centered and
scaled) data matrix and LASSO to generate sparsity on the loading vectors.

\code{scale= TRUE} is highly recommended as it will help obtaining
orthogonal sparse loading vectors.

\code{keepX} is the number of variables to keep in loading vectors. The
difference between number of columns of \code{X} and \code{keepX} is the
degree of sparsity, which refers to the number of zeros in each loading
vector.

Note that \code{spca} does not apply to the data matrix with missing values.
The biplot function for \code{spca} is not available.

According to Filzmoser et al., a ILR log ratio transformation is more
appropriate for PCA with compositional data. Both CLR and ILR are valid.

Logratio transform and multilevel analysis are performed sequentially as
internal pre-processing step, through \code{\link{logratio.transfo}} and
\code{\link{withinVariation}} respectively.

Logratio can only be applied if the data do not contain any 0 value (for
count data, we thus advise the normalise raw data with a 1 offset). For ILR
transformation and additional offset might be needed.
}
\examples{
#' \dontrun{
## successful: TRUE

library(mixOmics.data)
spca.rat <- spca(liver.toxicity$gene, ncomp = 3, keepX = rep(50, 3))
spca.rat

## variable representation
plotVar(spca.rat, cex=2)

plotVar(spca.rat,style="3d")
# }

# example with MultiAssayExperiment class
# --------------------------------

spca.rat <- spca(liver.toxicity.mae, assay='gene', ncomp = 3, keepX = rep(50, 3))
spca.rat

# \dontrun{


## samples representation
plotIndiv(spca.rat, ind.names = liver.toxicity$treatment[, 3],
          group = as.numeric(liver.toxicity$treatment[, 3]))
plotIndiv(spca.rat, cex = 0.01,
          col = as.numeric(liver.toxicity$treatment[, 3]),style="3d")


# example with multilevel decomposition and CLR log ratio transformation
# ----------------

data("diverse.16S")
pca.res = pca(X = diverse.16S$data.TSS, ncomp = 5,
              logratio = 'CLR', multilevel = diverse.16S$sample)
plot(pca.res)
plotIndiv(pca.res, ind.names = FALSE, group = diverse.16S$bodysite, title = '16S diverse data',
          legend=TRUE)

  #' }
}
\seealso{
\code{\link{pca}}, \code{\link{ipca}}, \code{\link{selectVar}},
\code{\link{plotIndiv}}, \code{\link{plotVar}} and http://www.mixOmics.org
for more details.
}
\author{
Ignacio Gonzalez, Kim-Anh LÃª Cao, Fangzhou Yao, Al J Abadi
}
\keyword{algebra}
